port/win/io_win.cc

```diff
@@ -183,25 +183,17 @@ size_t GetUniqueIdFromFile(HANDLE hFile, char* id, size_t max_size) {
 #if (_WIN32_WINNT == _WIN32_WINNT_VISTA)
   // MINGGW as defined by CMake file.
   // yuslepukhin: I hate the guts of the above macros.
-  // This impl does not guarantee uniqueness everywhere
-  // is reasonably good
-  BY_HANDLE_FILE_INFORMATION FileInfo;
 
-  BOOL result = GetFileInformationByHandle(hFile, &FileInfo);
-
-  TEST_SYNC_POINT_CALLBACK("GetUniqueIdFromFile:FS_IOC_GETVERSION", &result);
-
-  if (!result) {
-    return 0;
-  }
-
-  char* rid = id;
-  rid = EncodeVarint64(rid, uint64_t(FileInfo.dwVolumeSerialNumber));
-  rid = EncodeVarint64(rid, uint64_t(FileInfo.nFileIndexHigh));
-  rid = EncodeVarint64(rid, uint64_t(FileInfo.nFileIndexLow));
-
-  assert(rid >= id);
-  return static_cast<size_t>(rid - id);
+  // `FileIdInfo` is not supported. The possible alternative,
+  // `BY_HANDLE_FILE_INFORMATION::nFileIndex{High,Low}`, allows deleted files'
+  // IDs to be reused for newly created files. Since we don't evict from block
+  // cache before deleting a file, this introduces a risk that readers access
+  // obsolete data blocks.
+  //
+  // Returning 0 is safe as it causes the table reader to generate a unique ID.
+  // This is suboptimal for performance as it prevents multiple table readers
+  // for the same file from sharing cached blocks, but at least it's safe.
+  return 0;
```

Comment:
Isn't the unique-ID problem independent of MINGW. I was imagining we'd change `GetUniqueIdFromFile` to always return 0 on Windows, regardless of how the binary was built.

Reply:
I don't know. non-MinGW on a modern platform should have `_WIN32_WINNT != _WIN32_WINNT_VISTA` which means it uses the code in the `#else` below. I read a bit about the `FILE_ID_INFO::FileId` used there and could not determine whether it is reusable or not. At least there wasn't documentation explicitly stating IDs can be reused like there was for `BY_HANDLE_FILE_INFORMATION::nFileIndex{High,Low}`

Reply:
MinGW always uses this branch of the conditional compilation due to https://github.com/facebook/rocksdb/blob/2367656b6c19048d76037d24025ef0caab136866/CMakeLists.txt#L414-L416.

Reply:
https://stackoverflow.com/questions/1866454/unique-file-identifier-in-windows

> Support for file IDs is file system-specific. File IDs are not guaranteed to be unique over time, because file systems are free to reuse them. In some cases, the file ID for a file can change over time.
>
> In the FAT file system, the file ID is generated from the first cluster of the containing directory and the byte offset within the directory of the entry for the file. Some defragmentation products change this byte offset. (Windows in-box defragmentation does not.) Thus, a FAT file ID can change over time. Renaming a file in the FAT file system can also change the file ID, but only if the new file name is longer than the old one.

This all suggests to me that the problem is not MinGW specific.

Reply:
I don't know which file ID they're referring to. The one here is eight bytes, we know it has a problem, and it's only used when compiler is MinGW (or if somehow somebody compiles on Vista but I doubt that'll happen). The one below is sixteen bytes, we don't know if it's problematic, and it'll be used on modern non-MinGW builds.

Reply:
I'll find somebody from Microsoft to help. It's hard to believe it's broken beyond the case of `build for Vista || obsolete file system`, but who knows.

Reply:
It's still unclear so I'll `return 0` in all cases Windows to get this landed to our fork (note this makes no difference compared to the original PR for cockroach's use case, and I do not think these changes should be upstreamed).